rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md

```diff
@@ -0,0 +1,93 @@
+
+## Messages
+
+### SetFilter
+
+Upon receiving a `SetFilter` message, the remote peer will immediately restrict the broadcast transactions it announces to transactions matching the filter, where the [matching algorithm](#filter-matching-algorithm) is specified below.
+
+```
+table SetFilter {
+    filter: [uint8];
+    num_hashes: uint8;
+    hash_seed: uint32;
+}
+```
+
+`filter`: A bit field of arbitrary byte-aligned size. The maximum size is 36,000 bytes.
+
+`num_hashes`: The number of hash functions to use in this filter. The maximum value allowed in this field is 20. This maximum value and `filter` maximum size allow to store ~10,000 items and the false positive rate is 0.0001%.
+
+`hash_seed`: We use [Kirsch-Mitzenmacher-Optimization](https://www.eecs.harvard.edu/~michaelm/postscripts/tr-02-05.pdf) hash function in this protocol, `hash_seed` is a random offset, `h1` is low uint32 of hash value, `h2` is high uint32 of hash value, and the nth hash value is `(hash_seed + h1 + n * h2) mod filter_size`.
+
+### AddFilter
+
+Upon receiving a `AddFilter` message, the given bit data will be added to the exsiting filter via bitwise OR operator. A filter must have been previously provided using `SetFilter`. This messsage is useful if a new filter is added to a peer whilst it has connections to the network open, it avoids the need to re-calculate and send an entirely new filter to every peer.
+
+```
+table AddFilter {
+    filter: [uint8];
+}
+```
+
+`filter`: A bit field of arbitrary byte-aligned size. The data size must be litter than or equal to previously provided filter size.
+
+### ClearFilter
+
+The `ClearFilter` message tells the receiving peer to remove a previously-set bloom filter.
+
+```
+table ClearFilter {
+}
+```
+
+The `ClearFilter` message has no arguments at all.
+
+
+### FilteredBlock
+
+After a filter has been set, peers don't merely stop announcing non-matching transactions, they can also serve filtered blocks. This message is a replacement for `Block` message of sync protocol and `CompactBlock` message of relay protocol.
+
+```
+table FilteredBlock {
+    header: Header;
+    transactions: [[Transaction, uint32]];
+    hashes: [H256];
+}
+```
+
+`header`: Standard block header struct.
+
+`transactions`: Standard transaction struct plus transaction index.
+
+`hashes`: Partial Merkle branch proof.
+
+## Filter matching algorithm
+
+The filter can be tested against all broadcast transactions, to determine if a transaction matches the filter, the following algorithm is used. Once a match is found the algorithm aborts.
+
+1. Test the hash of the transaction itself.
+2. For each CellInput, test the hash of `previous_output` and `type hash` of unlock script.
+3. For each CellOutput, test the `lock` and `type hash` of contract.
```

Comment:
@xxuejie plz take a look at this

Reply:
Those seem a good list to me. One additional question is: is it true that we only implement a filter that matches a transaction as long as any of the fields here(such as `type hash`, `lock`, etc) matches the filter? In other words, will we implement some sort of filter that can just test for a single specified field, such as `lock`?

Reply:
> Those seem a good list to me. One additional question is: is it true that we only implement a filter that matches a transaction as long as any of the fields here(such as `type hash`, `lock`, etc) matches the filter? In other words, will we implement some sort of filter that can just test for a single specified field, such as `lock`?

Bloom filter is a probabilistic data structure, peer needs to check or filter transaction again after receiving it, so let's keep protocol simple.